### ISSUE TITLE
Revamp perf tests for service invocation

### ISSUE THREAD

**Issue by ItalyPaleAle** [dapr/dapr#6602]
We need to look into improving our perf tests for service invocation, both HTTP and gRPC, to address a number of issues:

- For now, the new tests should run **alongside** the old ones and not replace them. This is because we know the baselines using the old tests only. We can deprecate the old tests in the future.
- Switch the test framework to K6:
  - The current tests use fortio which doesn't support HTTP/2 (see #6601) and requires us to use a fork to support gRPC.
  - Current tests provide results that [aren't always statistically significant](https://github.com/dapr/dapr/pull/4782) due to how we handle computing baselines (in short, we first compute p50/p75/p90 of "baseline latency" which involves not using Dapr, then we run tests with Dapr and compute p50/p75/p90 latency, then we subtract the latency at each level to determine the "added latency" - e.g.. subtract p90 base latency from p90 total latency–, which is not correct and can lead to results such as negative added latency)
- The current perf tests perform invocations without any data in the body (of the request or response).
  - We should run tests with no body, with a small body (for example 1KB) and with a larger body (for example 1MB)
- Ensure that the pods (tester and test app) are always colocated on the same physical node, to remove concerns with data going over the network
- Consider switching the target app, currently written in Go, to a separate language such as Rust. The reason is that Go uses a garbage collection and in environments like our perf tests, where we fire thousands of requests per second, it could negatively impact the results of the tests.

**Comment by yaron2** [comment]
> The current perf tests perform invocations without any data in the body

Incorrect, tests run with [1Kb payload of the request](https://github.com/dapr/dapr/blob/1f88c38b2c9993d74dc27adcc8c9c3e1f1e51ac5/tests/perf/service_invocation_http/service_invocation_http_test.go#L97).

Switching to a framework that supports HTTP/2 would be a good change.

> then we subtract the latency at each level to determine the "added latency" - e.g.. subtract p90 base latency from p90 total latency–, which is not correct and can lead to results such as negative added latency

This method was modeled after the Istio and Linkerd performance tests to measure added latency. What would be the alternative way to measure added latency? In addition, can you find an example of a test run that shows negative added latency? This is interesting from a technical perspective.

**Comment by ItalyPaleAle** [comment]
> Incorrect, tests run with [1Kb payload of the request](https://github.com/dapr/dapr/blob/1f88c38b2c9993d74dc27adcc8c9c3e1f1e51ac5/tests/perf/service_invocation_http/service_invocation_http_test.go#L97).

Thanks for clarifying. It would still be nice to have tests with no body, small body (1KB is fine; as long as it's [less than 2KB](https://github.com/dapr/dapr/blob/1f88c38b2c9993d74dc27adcc8c9c3e1f1e51ac5/pkg/messaging/v1/util.go#L39-L40)), and larger body (e.g. 1MB).

>  In addition, can you find an example of a test run that shows negative added latency? This is interesting from a technical perspective.

I spent a bunch of time yesterday going through the messages in the repo but I couldn't find it anymore. I definitely do remember seeing negative added latency in some instances, especially before we removed AZs and started adding pod affinity rules.

Even when the added latency wasn't negative, we have seen situations where the added latency for example at p75 was < than added latency at p50. From the issue linked above:

![image](https://github.com/dapr/dapr/assets/43508/7380ad29-67f2-4e6e-a72e-ee4143a02ab2)


> What would be the alternative way to measure added latency?

The problem right now is that we subtract baseline latency from total latency at each percentile. For example, assume we have this data:

|     | Baseline latency | Total latency | Added latency |
|-----|------------------|---------------|---------------|
| p50 | 1.00ms           | 1.50ms        | 0.50ms        |
| p75 | 1.50ms           | 1.75ms        | 0.25ms        |
| p90 | 1.60ms           | 2.00ms        | 0.40ms        |

This data is fictitious but shows a similar situation to what we observed for real above.

- In the baseline latency, latency increases at p50, p75, and p90
- Same for total latency
- However, the "added latency" isn't increasing linearly

Why does that happen?

- We measure the baseline latency separately from the total latency, so the two measurements are not statistically correlated.
- When we subtract baseline latency from total latency at each percentile, we are acting as if the two measurements were correlated.

In short, just because 90% of baseline requests completed in less than 1.60ms, does not mean that the baseline latency for 90% of requests via Dapr is 1.60ms, as the two things are unrelated.

Statistics was never my favorite topic, so I may be completely off here, but I believe that it would be more accurate to subtract either median (p50) or average baseline latency from all total latency measurements.

**Comment by yaron2** [comment]
> Thanks for clarifying. It would still be nice to have tests with no body, small body (1KB is fine; as long as it's [less than 2KB](https://github.com/dapr/dapr/blob/1f88c38b2c9993d74dc27adcc8c9c3e1f1e51ac5/pkg/messaging/v1/util.go#L39-L40)), and larger body (e.g. 1MB).

Agree

> EDIT: Although this sends 1KB in the request body, it seems that the response body is always empty. Please confirm

That's correct. It'd be easy to enable the same data to be sent back via the response.
